### ISSUE TITLE
enumerate causes incompatible type mypy error

### ISSUE THREAD

**Issue by jroberts07** [python/mypy#11934]
**Bug Report**

When using the enumerate function mypy seems to generalise the output to `object` instead of respecting the type of the input

**To Reproduce**

```
from typing import Union


def process(actions: Union[list[str], list[int]]) -> None:
    for pos, action in enumerate(actions):
        act(action)


def act(action: Union[str, int]) -> None:
    print(action)
```

**Expected Behavior**

Mypy raises no errors

**Actual Behavior**

Mypy raises the following error: `Argument 1 to "act" has incompatible type "object"; expected "Union[str, int]"`

Its worth noting that the same code without using enumerate does not raise this error:

```
from typing import Union


def process(actions: Union[list[str], list[int]]) -> None:
    for action in actions:
        act(action)


def act(action: Union[str, int]) -> None:
    print(action)
```

**Your Environment**

- Mypy version used: 0.921
- Python version used: 0.9

**Comment by sobolevn** [comment]
You can try this (if you are using `python>=3.9`):

```python
from typing import Union

def process(actions: Union[list[str], list[int]]) -> None:
    for pos, action in enumerate[str | int](actions):
        act(action)

def act(action: Union[str, int]) -> None:
    print(action)
```

But, I agree that automatically infering `str | int` here can be a good feature. But, it is really hard to do 🙂

**Comment by erictraut** [comment]
This is yet another case where mypy's choice to use join rather than union causes problems. Pyright uses only unions, never joins, so it produces the expected result in this case.

Has there been any serious discussion of changing mypy's behavior to use unions instead of joins? I can't think of any case where joins are preferable, and there are many cases where unions are better because they retain more detailed type information.

**Comment by sobolevn** [comment]
CC @juk
